perun-core/src/main/java/cz/metacentrum/perun/core/impl/MembersManagerImpl.java

```diff
@@ -197,13 +198,24 @@
 
 		while(resultSet.next()) {
 			Member member = MembersManagerImpl.MEMBER_MAPPER_WITH_GROUP.mapRow(resultSet, resultSet.getRow());
+
 			if (member != null) {
+
+				GroupResourceStatus assignmentStatus = GroupResourceStatus.valueOf(resultSet.getString("group_resource_status"));
+
+				// if member is repeated, only update assignment status and richmember's group statuses
 				if (members.containsKey(member.getId())) {
-					members.get(member.getId()).getMember().putGroupStatuses(member.getGroupStatuses());
+					AssignedMember storedMember = members.get(member.getId());
+					storedMember.getRichMember().putGroupStatuses(member.getGroupStatuses());
+					if (assignmentStatus.isMoreImportantThan(storedMember.getStatus())) {
+						storedMember.setStatus(assignmentStatus);
```

Comment:
guys, I think I found the problem failing the tests randomly. I didn't update the assignment status of our member if he was already in the map structure and because memberId is used as key, he couldn't appear more times with different statuses which I relied on later, so the assignment status was always copied from his first entry in the database. I fixed it here and removed the iteration comparing the statuses that was used later.